# Assignment 3 – Tabular Dataset Preparation

This is Assignment 3 for the Introduction to Deep Learning with PyTorch course (www.leaky.ai).  In this assignment you will practice preparing tabular datasets for training a neural network.  You will practice applying normalization and standardization techniques.  You will also use pandas to convert categorical inputs into numerical values.
To Get Started

1.	Open up a web browser (preferable Chrome)
2.	Copy the Project GitHub Link: https://github.com/LeakyAI/PyTorch-Overview
3.	Head over to Google Colab (https://colab.research.google.com)
4.	Load the notebook: Tabular Dataset Preparation - Start Here.ipynb
5.	Replace the [TBD]'s with your own code
6.	Execute the notebook after completing each cell and check your answers using the solution notebook

Good Luck!

Key Takeaways:
•	You calculated the minimum and maximum values for each input and applied normalization
•	You then applied standardization and compared the results
•	You replaced categorical inputs with numerical values using one-hot encoding

## Part 1 - Standardization and Min Max Normalization

In [1]:
# Import PyTorch and set the seed for reproducible results
import torch
torch.set_printoptions(precision=3,sci_mode=False)  # Tensor easier to read

In [2]:
# Create a PyTorch tensor with the following content:
#    [[1,100,3,0.01,5000],[0,10,8,-0.002,0.01],[1,25,13,0.04,0.2],[1,45,18,-0.05,0.5]]
data = torch.tensor([[1,100,3,0.01,5000],[0,10,8,-0.002,0.01],[1,25,13,0.04,0.2],[1,45,18,-0.05,0.5]], dtype=torch.float)
print (data)

tensor([[     1.000,    100.000,      3.000,      0.010,   5000.000],
        [     0.000,     10.000,      8.000,     -0.002,      0.010],
        [     1.000,     25.000,     13.000,      0.040,      0.200],
        [     1.000,     45.000,     18.000,     -0.050,      0.500]])


#### Answer
<pre>
tensor([[     1.000,    100.000,      3.000,      0.010,   5000.000],
        [     0.000,     10.000,      8.000,     -0.002,      0.010],
        [     1.000,     25.000,     13.000,      0.040,      0.200],
        [     1.000,     45.000,     18.000,     -0.050,      0.500]])
</pre>

# Normalize the Values
Here you will apply normalization to the column values.

In [3]:
# Find the minimum and maximum value for each column
# Hint:  Make sure you use axis = 0 when calling min and max as we
#        want to apply the function calls to the columns (not entire tensor)
maximums = data.max(axis=0)
minimums = data.min(axis=0)
print (f"Max Values: {maximums.values}")
print (f"Min Values: {minimums.values}")

Max Values: tensor([    1.000,   100.000,    18.000,     0.040,  5000.000])
Min Values: tensor([     0.000,     10.000,      3.000,     -0.050,      0.010])


#### Answer
<pre>
Max Values: tensor([    1.000,   100.000,    18.000,     0.040,  5000.000])
Min Values: tensor([     0.000,     10.000,      3.000,     -0.050,      0.010])
</pre>

In [4]:
# Applying normalization to each input
# Use the formula x = (x-min)/(max-min)
dataNormalized = (data - minimums.values) / (maximums.values - minimums.values)
print (dataNormalized)

tensor([[    1.000,     1.000,     0.000,     0.667,     1.000],
        [    0.000,     0.000,     0.333,     0.533,     0.000],
        [    1.000,     0.167,     0.667,     1.000,     0.000],
        [    1.000,     0.389,     1.000,     0.000,     0.000]])


#### Answer
<pre>
tensor([[    1.000,     1.000,     0.000,     0.667,     1.000],
        [    0.000,     0.000,     0.333,     0.533,     0.000],
        [    1.000,     0.167,     0.667,     1.000,     0.000],
        [    1.000,     0.389,     1.000,     0.000,     0.000]])
</pre>

### Question
What obervations can be made about using normalization?  Does normalization work well in all cases?  How about the last column?

### Your Answer
Normalization ensures all our intputs are in the range of 0..1.  However, when the data contains outliers, as with the last column, normalization is not ideal as the smaller values will all be close to 0.

## Standardize the Values
Use the following formula:
xStandardized = (x - xMean) / xStdDeviation

In [5]:
# Calculate the mean and standard deviation of each column
dataMean = data.mean(axis = 0)
dataStDev = data.std(axis = 0)

print (f"Mean  : {dataMean}")
print (f"St Dev: {dataStDev}")

Mean  : tensor([     0.750,     45.000,     10.500,     -0.001,   1250.177])
St Dev: tensor([    0.500,    39.370,     6.455,     0.037,  2499.882])


#### Answer
<pre>
Mean  : tensor([     0.750,     45.000,     10.500,     -0.001,   1250.177])
St Dev: tensor([    0.500,    39.370,     6.455,     0.037,  2499.882])
</pre>

In [6]:
# Standadize the columns using the following formula:
# dataStandardized = (data - mean) / (standardDeviation)
# hint - make sure you use axis=0 as we want these operations
#        conducted on the columns (not rows, not entire tensor)
dataStandardized = (data - dataMean) / (dataStDev)
print (dataStandardized)

tensor([[ 0.500,  1.397, -1.162,  0.281,  1.500],
        [-1.500, -0.889, -0.387, -0.040, -0.500],
        [ 0.500, -0.508,  0.387,  1.082, -0.500],
        [ 0.500,  0.000,  1.162, -1.322, -0.500]])


#### Answer
<pre>
tensor([[ 0.500,  1.397, -1.162,  0.281,  1.500],
        [-1.500, -0.889, -0.387, -0.040, -0.500],
        [ 0.500, -0.508,  0.387,  1.082, -0.500],
        [ 0.500,  0.000,  1.162, -1.322, -0.500]])
</pre>

### Question
What obervations can be made about using normalization?  Does normalization work well in all cases?  How about the last column?

### Your Answer
Standardization resulted in a larger range (outside of 0..1) but also ensured that the outlier in the last column did not overly affect the other input values in the column.  For columns with outliers, it is usually better to apply standardization as opposed to normalization.


## One-Hot Encoding
Most tabular datasets contain categorical data.  You will need to convert this type of data into numerical data before training.  We will be using the panda library to automatically convert our the categorical data into numeric using the get_dummies function.

In [7]:
# Load a categorical dataset using Pandas
import pandas as pd
df = pd.read_csv('cat_data_v1.csv')

In [8]:
# Understand the shape of the data by displaying the value of shape
df.shape

(1000, 3)

In [9]:
# Show the first portion of the data using head()
df.head()

Day of the Week Raining           Age Group
0       Wednesday      No        10..65 years
1          Monday     Yes  More than 65 years
2          Monday     Yes  More than 65 years
3          Friday     Yes  Less than 10 years
4       Wednesday      No        10..65 years

In [10]:
# Use the describe() function to better understand the data
# and look for missing values
df.describe()

Day of the Week Raining           Age Group
count             1000     945                1000
unique               7       2                   3
top           Thursday      No  Less than 10 years
freq               155     496                 347

In [11]:
# Drop rows that contain missing values using dropna()
df = df.dropna()
df.describe()

Day of the Week Raining           Age Group
count              945     945                 945
unique               7       2                   3
top           Thursday      No  Less than 10 years
freq               149     496                 327

In [12]:
# Create one-hot encoded values for each column using
# the the get_dummies function:
OneHot = pd.get_dummies(df, drop_first=True)
OneHot.head()

Day of the Week_Monday  Day of the Week_Saturday  Day of the Week_Sunday  \
0                       0                         0                       0   
1                       1                         0                       0   
2                       1                         0                       0   
3                       0                         0                       0   
4                       0                         0                       0   

   Day of the Week_Thursday  Day of the Week_Tuesday  \
0                         0                        0   
1                         0                        0   
2                         0                        0   
3                         0                        0   
4                         0                        0   

   Day of the Week_Wednesday  Raining_Yes  Age Group_Less than 10 years  \
0                          1            0                             0   
1                          0            1                             0   
2                          0            1                             0   
3                          0            1                             1   
4                          1            0                             0   

   Age Group_More than 65 years  
0                             0  
1                             1  
2                             1  
3                             0  
4                             0

### Key Takeaways:
- You calculated the minimum and maximum values for each input and applied normalization
- You then applied standardization and compared the results
- You replaced categorical inputs with numerical values using one-hot encoding